gpii/node_modules/preferencesServer/src/preferencesService.js

```diff
@@ -73,12 +124,270 @@ var fluid = fluid || require("infusion");
+                           gpiiExpressUserId: "{arguments}.0._id"
+                       }],
+                namespace: "addRecord"
+            }, {
+                funcName: "{dataStore}.findById",
+                args: ["{arguments}.0.id"]
+            }],
+            onCloudSafeUnlock: [{
+                func: "{expressUserUtils}.unlockUser",
+                args: ["{arguments}.0.username", "{arguments}.0.password"], // username, password
+                namespace: "unlockUser"
+            }, {
+                func: "{dataStore}.findSafeByExpressUserLookup",
+                args: ["{arguments}.0._id"],
+                namespace: "findSafeByExpressUserLookup"
+            }, {
+                func: "{dataStore}.findById",
+                args: ["{arguments}.0.prefsSafeId"],
+                namespace: "findById"
             }]
         }
     });
 
     // APIs for Preferences and Preferences Safes
 
+    gpii.preferencesServer.preferencesService.errors = fluid.freezeRecursive({
+        gpiiKeyMissing: "GPII key \"%gpiiKey\" does not exist",
+        gpiiKeyExisted: "GPII key \"%gpiiKey\" already exists",
+        noUpdateOnSnapset: "Cannot update:  GPII key \"%gpiiKey\" is a snapset"
+    });
+
+    /**
+     * Returns a preferences safe using it's id. If the safe is not found an error object
+     * is returned.
+     *
+     * @param  {Object} dataStore - Instance of `gpii.dbOperation.dataStore`
+     * @param  {String} prefsSafeId Prefs Safe ID
+     * @return {Promise}            Returns a promise resolving with the preferences safe,
+     * or rejecting with an error payload containing an error `message`.
+     */
+    gpii.preferencesServer.preferencesService.getPrefsSafe = function (dataStore, prefsSafeId) {
+        var promiseTogo = fluid.promise();
+        dataStore.findById(prefsSafeId).then(
+            function (data) {
+                if ($.isEmptyObject(data) || data.type !== "prefsSafe") {
+                    promiseTogo.reject({
+                        isError: true,
+                        errorCode: gpii.preferencesServer.errors.missingPrefsSafe.errorCode,
+                        message: gpii.preferencesServer.errors.missingPrefsSafe.message
+                    });
+                }
+                else {
+                    promiseTogo.resolve(data);
+                }
+            },
+            function (err) {
+                fluid.log("Error looking up prefsSafe: ", err);
+                promiseTogo.reject({
+                    isError: true,
+                    errorCode: gpii.preferencesServer.errors.errorAccessingPrefsSafe.errorCode,
+                    message: gpii.preferencesServer.errors.errorAccessingPrefsSafe.message
+                });
+            }
+        );
+        return promiseTogo;
+    };
+
+    /**
+     * Adds a new preferences safe to CouchDB using the preferences safe format
+     * minus the `id`, `timestampCreated`, and `timestampUpdated` properties.
```

Comment:
This statement of "minus the `id`, `timestampCreated`, and `timestampUpdated` properties" is misleading. It was the input `prefsSafeData` value does not require these fields, but this function will end up with creating all these fields in the database. Probably use this statement for `@param {Object} prefsSafeData` and re-describe the function goal here.

Reply:
updated